common/archiver/archivalMetadata.go

```diff
@@ -73,25 +74,70 @@ const (
 
 // NewArchivalMetadata constructs a new ArchivalMetadata
 func NewArchivalMetadata(
+	dc *dynamicconfig.Collection,
 	historyStatus string,
 	historyReadEnabled bool,
 	visibilityStatus string,
 	visibilityReadEnabled bool,
 	domainDefaults *config.ArchivalDomainDefaults,
 ) ArchivalMetadata {
+	historyConfig := NewArchivalConfig(
+		historyStatus,
+		historyReadEnabled,
+		domainDefaults.History.Status,
+		domainDefaults.History.URI,
+	)
+	if historyConfig.ClusterConfiguredForArchival() {
+		// Only check dynamic config when archival is enabled in static config.
+		// Dynamic archival config is accessed once on cluster startup than never accessed again.
+		// This is done so as to keep archival status and and the initialization of archiver.Provider in sync.
+		// TODO: Once archival pause is implemented archival config can be made truly dynamic.
+		dynamicHistoryStatusString := dc.GetStringProperty(dynamicconfig.HistoryArchivalStatus, historyStatus)()
+		dynamicHistoryStatus, err := getClusterArchivalStatus(dynamicHistoryStatusString)
+		if err != nil {
+			panic(err)
+		}
+		if dynamicHistoryStatus != ArchivalEnabled {
+			// apply status override
+			historyConfig = NewDisabledArchvialConfig()
+		} else {
+			// apply read override
+			historyConfig = NewArchivalConfig(
+				dynamicHistoryStatusString,
+				dc.GetBoolProperty(dynamicconfig.EnableReadFromHistoryArchival, historyReadEnabled)(),
+				domainDefaults.History.Status,
+				domainDefaults.History.URI,
+			)
+		}
+	}
+
+	visibilityConfig := NewArchivalConfig(
+		visibilityStatus,
+		visibilityReadEnabled,
+		domainDefaults.Visibility.Status,
+		domainDefaults.Visibility.URI,
+	)
+	if visibilityConfig.ClusterConfiguredForArchival() {
+		dynamicVisibilityStatusString := dc.GetStringProperty(dynamicconfig.VisibilityArchivalStatus, visibilityStatus)()
+		dynamicVisibilityStatus, err := getClusterArchivalStatus(dynamicVisibilityStatusString)
+		if err != nil {
+			panic(err)
+		}
+		if dynamicVisibilityStatus != ArchivalEnabled {
+			visibilityConfig = NewDisabledArchvialConfig()
+		} else {
+			visibilityConfig = NewArchivalConfig(
+				dynamicVisibilityStatusString,
+				dc.GetBoolProperty(dynamicconfig.EnableReadFromVisibilityArchival, visibilityReadEnabled)(),
+				domainDefaults.Visibility.Status,
+				domainDefaults.Visibility.URI,
+			)
+		}
+	}
+
 	return &archivalMetadata{
-		historyConfig: NewArchivalConfig(
-			historyStatus,
-			historyReadEnabled,
-			domainDefaults.History.Status,
-			domainDefaults.History.URI,
-		),
-		visibilityConfig: NewArchivalConfig(
-			visibilityStatus,
-			visibilityReadEnabled,
-			domainDefaults.Visibility.Status,
-			domainDefaults.Visibility.URI,
-		),
+		historyConfig:    historyConfig,
+		visibilityConfig: visibilityConfig,
```

Comment:
What do you think about just setting visibilityConfig: NewDisabledArchivalConfg()

The advantage of this I see is it becomes really obvious that visibility archival cannot be enabled